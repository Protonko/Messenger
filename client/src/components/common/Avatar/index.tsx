// types
import {FC} from 'react'
import {Sizes} from 'models/common/sizes'

import {Loader} from 'components/common/Loader'

type TAvatarClassNames =
  | 'avatar'
  | 'avatar avatar--sm'
  | 'avatar avatar--lg'

export interface IPropsAvatar {
  src: string,
  size?: Sizes,
}

export const Avatar: FC<IPropsAvatar> = ({
  src,
  size = Sizes.MEDIUM,
}) => {
  let avatarClassname: TAvatarClassNames;

  switch (size) {
      case Sizes.SMALL:
        avatarClassname = 'avatar avatar--sm'
        break
      case Sizes.LARGE:
        avatarClassname = 'avatar avatar--lg'
        break
    default:
      avatarClassname = 'avatar'
  }

  return (
    <div className={avatarClassname}>
      <img className="avatar__image" src={src} alt=""/>
      <Loader />
    </div>
  )
}
