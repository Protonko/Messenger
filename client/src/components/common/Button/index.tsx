import {FC, HTMLProps, ReactNode} from 'react'

interface IButtonProps extends HTMLProps<HTMLButtonElement> {
  text?: string
  icon?: ReactNode
  leftAddons?: ReactNode
  rightAddons?: ReactNode
  type?: 'button' | 'submit' | 'reset'
}

export const Button: FC<IButtonProps> = ({
  text,
  icon,
  leftAddons,
  rightAddons,
  ...buttonProps
}) => {
  const renderButtonContent = () => {
    return (
      <>
        {text && (
          <span className="button__text">
            {text}
          </span>
        )}
        {icon && (
          <span className="button__icon">
            {icon}
          </span>
        )}
      </>
    )
  }

  return (
    <button className="button" {...buttonProps}>
      {leftAddons}
      {renderButtonContent()}
      {rightAddons}
    </button>
  )
}
