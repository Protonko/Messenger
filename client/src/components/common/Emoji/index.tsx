import 'emoji-mart/css/emoji-mart.css'
import {FC, useState, useRef} from 'react'
import {EmojiData, Picker} from 'emoji-mart'
import {ReactComponent as Smile} from 'assets/icons/smile.svg'
import {useOutsideClick} from 'hooks/useOutsideClick'

export interface IPropsEmoji {
  onSelect: (emoji: EmojiData) => void
}

export const Emoji: FC<IPropsEmoji> = ({onSelect}) => {
  const [open, setOpen] = useState(false)
  const ref = useRef(null)
  const initiatorRef = useRef(null)

  useOutsideClick(ref, () => setOpen(false), initiatorRef)

  const onSelectEmoji = (emoji: EmojiData) => {
    onSelect(emoji)
    setOpen(false)
  }

  return (
    <div className="emoji">
      {open && (
        <div className="emoji__popup" ref={ref}>
          <Picker native={true} onSelect={onSelectEmoji} />
        </div>
      )}

      <Smile
        className="emoji__button"
        ref={initiatorRef}
        onClick={() => setOpen(!open)}
      />
    </div>
  )
}
