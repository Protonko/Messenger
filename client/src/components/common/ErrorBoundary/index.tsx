import {Component, ErrorInfo} from 'react';

interface IProps {}
interface IState {
  error: null | Error;
  errorInfo: null | ErrorInfo;
  hasError: boolean;
}

export class ErrorBoundary extends Component<IProps, IState> {
  constructor(props: IProps) {
    super(props);

    this.state = {
      error: null,
      errorInfo: null,
      hasError: false,
    }
  }

  static getDerivedStateFromError(error: Error) {
    console.log(error)
    return {
      hasError: true,
    };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    })
  }

  _renderDetailInfo() {
    if (process.env.NODE_ENV === 'development') {
      return (
        <div>

        </div>
      )
    }

    return null
  }

  render() {
    if (this.state.hasError) {
      return (
        <>
          <h1>Что-то пошло не так.</h1>

          {process.env.NODE_ENV === 'development' && (
            <details style={{ whiteSpace: 'pre-wrap' }}>
              {this.state.error && this.state.error.toString()}
              <br />
              {this.state.errorInfo?.componentStack}
            </details>
          )}
        </>
      )
    }

    return this.props.children;
  }
}
