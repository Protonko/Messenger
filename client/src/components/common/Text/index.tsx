import type {FC} from 'react'
import {
  TextTypes,
  TextSize,
  TextWeight,
  TNumberOfLines,
} from 'models/common/text'
import classNames from 'classnames'
import {classNamesText} from 'utils/classNamesText'
import {withDot} from 'hoc/withDot'

export interface IPropsText {
  type?: TextTypes
  weight?: TextWeight
  size?: TextSize
  children: string
  customStyles?: string
  numberOfLines?: TNumberOfLines
}

export const Text: FC<IPropsText> = withDot<IPropsText>(
  ({type = TextTypes.p, weight, size, children, customStyles}) => {
    const classNamesMixed = classNames(
      'text',
      // weight
      {'text--light': weight === TextWeight.LIGHT},
      {'text--regular': weight === TextWeight.REGULAR},
      {'text--medium': weight === TextWeight.MEDIUM},
      {'text--bold': weight === TextWeight.BOLD},
      // sizes
      {'text--xs': size === TextSize.EXTRA_SMALL},
      {'text--sm': size === TextSize.SMALL},
      {'text--md': size === TextSize.MEDIUM},
      {'text--lg': size === TextSize.BIG},
      {'text--xl': size === TextSize.LARGE},
      {'text--xxl': size === TextSize.EXTRA_LARGE},
      // custom
      {[customStyles ?? '']: !!customStyles},
    )

    switch (type) {
      case TextTypes.h1:
        return (
          <h1 className={classNamesText('heading-1', customStyles)}>
            {children}
          </h1>
        )

      case TextTypes.h2:
        return (
          <h2 className={classNamesText('heading-2', customStyles)}>
            {children}
          </h2>
        )

      case TextTypes.h3:
        return (
          <h3 className={classNamesText('heading-3', customStyles)}>
            {children}
          </h3>
        )

      case TextTypes.h4:
        return (
          <h4 className={classNamesText('heading-4', customStyles)}>
            {children}
          </h4>
        )

      case TextTypes.h5:
        return (
          <h5 className={classNamesText('heading-5', customStyles)}>
            {children}
          </h5>
        )

      case TextTypes.h6:
        return (
          <h6 className={classNamesText('heading-6', customStyles)}>
            {children}
          </h6>
        )

      case TextTypes.p:
        return (
          <p className={classNamesText('description', customStyles)}>
            {children}
          </p>
        )

      case TextTypes.mixed:
        return <p className={classNamesMixed}>{children}</p>

      default:
        return (
          <p className={classNamesText('description', customStyles)}>
            {children}
          </p>
        )
    }
  },
)
