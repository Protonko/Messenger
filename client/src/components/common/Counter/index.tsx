// types
import {Status} from 'models/common/status'

import {FC} from 'react'
import classNames from 'classnames'

export interface IPropsCounter {
  count: number | '99+',
  status?: Status,
}

export const Counter: FC<IPropsCounter> = ({
  count,
  status = Status.ACTIVE,
}) => {
  const classNameStatus = classNames(
    'counter',
    {'counter--muted': status === Status.MUTED}
  )

  return (
    <div className={classNameStatus}>
      <span className="counter__text">
        {count}
      </span>
    </div>
  )
}
