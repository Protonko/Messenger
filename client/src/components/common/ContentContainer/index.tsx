import type {FC} from 'react'
import {Loader} from 'components/common/Loader'
import {Text} from 'components/common/Text'

interface ContentContainerProps {
  loading: boolean
  errorMessage: string | null
  children: JSX.Element
}

export const ContentContainer: FC<ContentContainerProps> = ({
  loading,
  errorMessage,
  children,
}) => {
  if (errorMessage) {
    return <Text>{errorMessage}</Text>
  }

  if (loading) {
    return <Loader center={true} />
  }

  return children
}
