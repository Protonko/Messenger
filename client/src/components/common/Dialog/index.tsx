// types
import {TextSize, TextTypes} from 'models/common/text';
import {IDialog} from 'models/dialog';

import {FC} from 'react'
import {Text} from 'components/common/Text';
import {Counter} from 'components/common/Counter';
import {Avatar} from 'components/common/Avatar';

interface IPropsDialog extends IDialog {}

export const Dialog: FC<IPropsDialog> = ({
  name,
  description,
  avatar,
  date,
  time,
  messages,
  status,
  readStatus,
}) => {
  const dateTime = time ?? date;
  const counter = messages.toString().length > 2 ? '99+' : messages;

  return (
    <div className="dialog">
      <Avatar src={avatar} />

      <div className="dialog__data">
        <div className="dialog__text">
          <Text
            type={TextTypes.h4}
            customStyles="dialog__text-title"
            numberOfLines={1}
          >
            {name}
          </Text>
          <Text>{description}</Text>
        </div>

        <div className="dialog__info">
          <Text
            type={TextTypes.mixed}
            size={TextSize.EXTRA_SMALL}
            customStyles="dialog__info-date"
          >
            {dateTime}
          </Text>

          {!!messages ? <Counter count={counter} status={status}/> : null}
        </div>
      </div>
    </div>
  )
}
