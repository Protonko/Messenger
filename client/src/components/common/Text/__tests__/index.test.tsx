import {render} from 'enzyme'
import {IPropsText, Text} from 'components/common/Text'
import {TextTypes, TextWeight} from 'models/common/text'

const ELEMENT_SELECTORS = {
  text: 'text',
  paragraph: 'p',
  heading1: 'text--heading-1',
  heading2: 'text--heading-2',
  heading3: 'text--heading-3',
  heading4: 'text--heading-4',
  heading5: 'text--heading-5',
  heading6: 'text--heading-6',
  light: 'text--light',
  custom: 'custom-style',
}

const renderComponent = (props: IPropsText) => render(<Text {...props} />)

describe('Text', () => {
  let props: IPropsText

  beforeEach(() => {
    props = {
      children: 'Hello world',
    }
  })

  it('Should render Text component', () => {
    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.text}`)

    expect(wrapper.length).toBe(1)
  })

  it('Match snapshot', () => {
    const component = renderComponent(props)

    expect(component).toMatchSnapshot()
  })

  it('Should render <p></p>', () => {
    const component = renderComponent(props)
    const wrapper = component.find(ELEMENT_SELECTORS.paragraph)

    expect(wrapper.length).toBe(1)
  })

  it('Should render heading-1', () => {
    props.type = TextTypes.h1

    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.heading1}`)

    expect(wrapper.length).toBe(1)
  })

  it('Should render heading-2', () => {
    props.type = TextTypes.h2

    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.heading2}`)

    expect(wrapper.length).toBe(1)
  })

  it('Should render heading-3', () => {
    props.type = TextTypes.h3

    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.heading3}`)

    expect(wrapper.length).toBe(1)
  })

  it('Should render heading-4', () => {
    props.type = TextTypes.h4

    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.heading4}`)

    expect(wrapper.length).toBe(1)
  })

  it('Should render heading-5', () => {
    props.type = TextTypes.h5

    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.heading5}`)

    expect(wrapper.length).toBe(1)
  })

  it('Should render heading-6', () => {
    props.type = TextTypes.h6

    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.heading6}`)

    expect(wrapper.length).toBe(1)
  })

  it('Should render <p></p> as default value', () => {
    // @ts-ignore
    props.type = 'another type'

    const component = renderComponent(props)
    const wrapper = component.find(ELEMENT_SELECTORS.paragraph)

    expect(wrapper.length).toBe(1)
  })

  it('Should render light font weight', () => {
    props.type = TextTypes.mixed
    props.weight = TextWeight.LIGHT

    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.light}`)

    expect(wrapper.length).toBe(1)
  })

  it('Should render custom styles', () => {
    props.customStyles = ELEMENT_SELECTORS.custom

    const component = renderComponent(props)
    const wrapper = component.find(`.${ELEMENT_SELECTORS.custom}`)

    expect(wrapper.length).toBe(1)
  })
})
