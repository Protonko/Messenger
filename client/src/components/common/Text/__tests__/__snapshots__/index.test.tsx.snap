// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Match snapshot 1`] = `
<div>
  <p
    class="text text--description"
  >
    Hello world
  </p>
</div>
`;
